### PR TITLE
Port code from CygwinCCompiler to UnixCCompiler

### PR DESCRIPTION
**Pull request by DWesl**
https://github.com/python-pillow/Pillow/issues/7158#issuecomment-1546746716 and 
https://github.com/pypa/setuptools/blob/e76db4408e40a5c0a218ef6f652454f69c5aca59/setuptools/_distutils/ccompiler.py#L1055-L1058 
suggest that Cygwin uses `UnixCCompiler` rather than `CygwinCCompiler` by default, so `UnixCCompiler` would also need to know how to find shared libraries, import libraries, and static libraries on Cygwin.

Port of #139 to `UnixCCompiler`

### CODE REVIEW

**Comment by DWesl** [comment]
`UnixCCompiler.find_library_file` now returns library names that are likely to occur on Cygwin, rather than those likely on Linux, which causes the doctest to fail.  Should I fix that with `# doctest: ELLIPSIS` or something like
```python
>>> result = compiler.find_library_file(dirs, 'abc').replace('\\', '/')
>>> if sys.platform == "cygwin":
...     print(result == "/foo/bar/existing/cygabc.dll")
... else:
...     print(result == "/foo/bar/existing/libabc.dylib")
True
```

**Comment by DWesl** [comment]
Given [Cygwin seems to use `UnixCCompiler` rather than `CygwinCCompiler`](https://github.com/pypa/setuptools/blob/e76db4408e40a5c0a218ef6f652454f69c5aca59/setuptools/_distutils/ccompiler.py#L1056-L1058), should we also drop `cygwinccompiler.py`?

**Comment by radarhere** [comment]
> Given [Cygwin seems to use `UnixCCompiler` rather than `CygwinCCompiler`](https://github.com/pypa/setuptools/blob/e76db4408e40a5c0a218ef6f652454f69c5aca59/setuptools/_distutils/ccompiler.py#L1056-L1058), should we also drop `cygwinccompiler.py`?

#185 similarly feels CygwinCCompiler isn't used, although thinks it should be.

**Comment by DWesl** [comment]
The odd doctests suggest having `CygwinCCompiler` be a thin wrapper around `UnixCCompiler` might be the way to go: the only thing I can think to change would be the doctests in this PR, as everything else seems fine.

File: distutils/unixccompiler.py [hunk 1]
```diff
@@ -141,6 +141,10 @@ class UnixCCompiler(CCompiler):
     xcode_stub_lib_format = dylib_lib_format
     if sys.platform == "cygwin":
         exe_extension = ".exe"
+        shared_lib_extension = ".dll.a"
+        dylib_lib_extension = ".dll"
+        static_lib_format = shared_lib_format = "lib%s%s"
```
Comment (DWesl) [inline distutils/unixccompiler.py:146]:
```suggestion
```
The pattern here is the same as on Linux, just with different extensions.  We can remove this and rely on the definition on line 140.

**Comment by jaraco** [comment]
Oh, heck. Tests on cygwin are currently disabled because ruff fails to install there.

**Comment by DWesl** [comment]
There's some progress on rust-lang/rust#79854; I'm somewhat curious how well the work done to cross-compile from Linux would carry over to cross-compiling from Windows.

**Comment by ookiineko** [comment]
>  I'm somewhat curious how well the work done to cross-compile from Linux would carry over to cross-compiling from Windows.

there wasn't any existing cygwin cross compiler available on mingw or msys, but there is [one on fedora linux](https://copr.fedorainfracloud.org/coprs/yselkowitz/cygwin/)

i also tried to get a cygwin cross compiler on archlinux: [arch-cygwin](https://github.com/ookiineko/arch-cygwin) (unofficial), it uses PKGBUILD (which is the same as what MSYS2 is using), might be useful if someone want to build them and package on msys2

**Comment by jaraco** [comment]
> How should I be fixing the doctests to not fail with working code on Cygwin?

I don't see any problem with the doctests. Perhaps they've been resolved.

The only failing test now is the failing coverage test, which is spurious and only there because the coverage is only checked on one platform, so it's ignorable.

**Comment by DWesl** [comment]
> > How should I be fixing the doctests to not fail with working code on Cygwin?
> 
> I don't see any problem with the doctests. Perhaps they've been resolved.

It was a failure on the Cygwin doctests, which I think got disabled because there's no Rust compiler for Cygwin

**Comment by Kreijstal** [comment]
> > > How should I be fixing the doctests to not fail with working code on Cygwin?
> > 
> > 
> > I don't see any problem with the doctests. Perhaps they've been resolved.
> 
> It was a failure on the Cygwin doctests, which I think got disabled because there's no Rust compiler for Cygwin

the problem is usually path conversion
```
https://github.com/thepowersgang/mrustc
```
you can kinda get rust here on cygwin, but there are many problems with path conversion and cygwin not being linux

**Pull request merged**